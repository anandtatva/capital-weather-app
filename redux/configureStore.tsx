import { createStore, combineReducers } from "redux";
import { persistReducer, persistStore } from "redux-persist"
import AsyncStorage from "@react-native-async-storage/async-storage"
import CountrySlice from "./reducer/country-slice"
const persistConfig = {
    key: "country",
    storage: AsyncStorage
}

const rootReducer = combineReducers({
    country: CountrySlice
})

const persistReducers = persistReducer(persistConfig, rootReducer)

const store = createStore(persistReducers);

const persistor = persistStore(store)

export {
    store,
    persistor
}