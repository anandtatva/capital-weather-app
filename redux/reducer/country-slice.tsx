import { createSlice } from "@reduxjs/toolkit";
import { CountryType } from "../../types/country-type"
import { CapitalWeatherType } from "../../types/capital-weather-type"


interface CountryState {
    name: string;
    countries: CountryType[];
    capitalWeather: CapitalWeatherType | null;
}

const initialState: CountryState = {
    name: "",
    capitalWeather: null,
    countries: []
}

const CountrySlice = createSlice({
    name: "country-slice",
    initialState,
    reducers: {
        setName: (state, action) => {
            state.name = action.payload
        },
        setCountries: (state, action) => {
            state.countries = action.payload
        },
        setCapitalWeather: (state, action) => {
            state.capitalWeather = action.payload
        },

    }
})


export const { setCapitalWeather, setCountries, setName } = CountrySlice.actions;
export default CountrySlice.reducer;