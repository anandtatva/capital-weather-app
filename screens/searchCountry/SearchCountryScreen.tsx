import { View, Text, TextInput, StyleSheet } from "react-native";
import { strings, colors, showAlert } from "../../utils";
import { useSelector, useDispatch } from "react-redux";
import { setName, setCountries } from "../../redux/reducer/country-slice";
import Button from "../../components/Button";
import { searchCountryAPIs } from "../../services/api";

const SearchCountryScreen = ({ navigation }) => {
    const country = useSelector(state => state.country);
    const dispatch = useDispatch();

    const handleChange = (text: string) => {
        dispatch(setName(text));
    }
    const handleSearch = async () => {
        const resp = await searchCountryAPIs(country.name)
        if (resp.status) {
            dispatch(setCountries(resp.data));
            navigation.navigate("CountryListScreen")
        } else showAlert(resp.message)
    }
    return (
        <View style={styles.container}>
            <TextInput
                style={styles.input}
                testID="Name-Input"
                placeholder={strings.countryPlaceholder}
                placeholderTextColor={colors.gray}
                value={country.name}
                onChangeText={handleChange}
            />
            <Button
                disabled={!country.name}
                name="Search Country"
                onButtonPress={handleSearch}
                testId={"Search-Country-Button"}
            />
        </View>
    )
}

export default SearchCountryScreen;

const styles = StyleSheet.create({
    container: {
        padding: 20
    },
    input: {
        paddingVertical: 5,
        paddingHorizontal: 10,
        borderWidth: 1,
        borderColor: colors.primary,
        borderRadius: 2
    }
})