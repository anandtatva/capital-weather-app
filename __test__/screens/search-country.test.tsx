import { render, waitFor, cleanup, fireEvent, act } from "@testing-library/react-native";
import { Provider } from "react-redux";
import { store } from "../../redux/configureStore";
import SearchCountryScreen from "../../screens/searchCountry/SearchCountryScreen";
import { setName, setCountries, setCapitalWeather } from "../../redux/reducer/country-slice";
import { Alert } from "react-native";

jest.mock("@react-navigation/native", () => {
    const actualNav = jest.requireActual("@react-navigation/native")
    return {
        ...actualNav,
        useFocusEffect: () => jest.fn(),
        useNavigation: () => ({
            navigate: jest.fn()
        }),
        useRoute: () => ({
            params: {}
        })
    }
})

const spyAlert = jest.spyOn(Alert, "alert")

describe("Screen : Search Country", () => {
    beforeEach(() => {
        fetchMock.mockIf(/^https?:\/\/example.com.*S/)
        store.dispatch(setName(""));
        store.dispatch(setCountries([]))
        store.dispatch(setCapitalWeather(null))
    })
    afterEach(() => {
        jest.clearAllMocks();
        cleanup()
    })
    it("renders correctly", () => {
        const navigation = { navigate: jest.fn() }
        const tree = render(
            <Provider store={store}>
                <SearchCountryScreen navigation={navigation} />
            </Provider>
        ).toJSON()
        expect(tree).toMatchSnapshot();
    })
    it("Button should be disabled if no input entered", () => {
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <SearchCountryScreen navigation={navigation} />
            </Provider>
        )
        const searchButton = getByTestId("Search-Country-Button");
        expect(searchButton.props.accessibilityState.disabled).toBeTruthy();
    })
    it("Button should be clickable as input enter", async () => {
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <SearchCountryScreen navigation={navigation} />
            </Provider>
        )
        const inputText = "india";
        const searchButton = getByTestId("Search-Country-Button");
        expect(searchButton.props.accessibilityState.disabled).toBeTruthy();

        const nameInput = getByTestId("Name-Input");
        await waitFor(() => {
            fireEvent.changeText(nameInput, inputText)
        })
        const searchButtonAfterChange = getByTestId("Search-Country-Button");
        expect(searchButtonAfterChange.props.accessibilityState.disabled).toBeFalsy();
    })
    it("Input value should be change", async () => {
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <SearchCountryScreen navigation={navigation} />
            </Provider>
        )
        const inputText = "india";
        const searchButton = getByTestId("Search-Country-Button");
        expect(searchButton.props.accessibilityState.disabled).toBeTruthy();
        const nameInput = getByTestId("Name-Input");
        await waitFor(() => {
            fireEvent.changeText(nameInput, inputText)
        })
        expect(store.getState().country.name).toBe(inputText)
    })
    it("Valid Countries should be find on valid name", async () => {
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <SearchCountryScreen navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            const inputText = "india";
            const nameInput = getByTestId("Name-Input");
            await waitFor(() => {
                fireEvent.changeText(nameInput, inputText)
            })
            const searchButton = getByTestId("Search-Country-Button");
            await fireEvent.press(searchButton)
            expect(store.getState().country.countries.length).toBe(2);
            expect(navigation.navigate).toHaveBeenCalledWith("CountryListScreen")
        })
    })
    it("Error should be show on invalid name", async () => {
        const navigation = { navigate: jest.fn() }
        const { getByTestId } = render(
            <Provider store={store}>
                <SearchCountryScreen navigation={navigation} />
            </Provider>
        )
        await act(async () => {
            const inputText = "AAAAA";
            const nameInput = getByTestId("Name-Input");
            await waitFor(() => {
                fireEvent.changeText(nameInput, inputText)
            })
            const searchButton = getByTestId("Search-Country-Button");
            await fireEvent.press(searchButton)
            expect(Alert.alert).toHaveBeenCalled();
            // @ts-ignore
            expect(spyAlert.mock.calls[0][0]).toBe("Country Not Found");
        })
    })
})