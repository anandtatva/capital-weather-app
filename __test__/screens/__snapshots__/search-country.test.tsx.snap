// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screen : Search Country renders correctly 1`] = `
<View
  style={
    Object {
      "padding": 20,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter Country Name"
    placeholderTextColor="gray"
    style={
      Object {
        "borderColor": "#000",
        "borderRadius": 2,
        "borderWidth": 1,
        "paddingHorizontal": 10,
        "paddingVertical": 5,
      }
    }
    testID="Name-Input"
    value=""
  />
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#000",
        "marginVertical": 20,
        "opacity": 0.8,
        "paddingVertical": 10,
      }
    }
    testID="Search-Country-Button"
  >
    <Text
      style={
        Array [
          Object {
            "color": "#FFF",
            "textAlign": "center",
          },
          Object {},
        ]
      }
    >
      Search Country
    </Text>
  </View>
</View>
`;
